src/apiutil.h

```diff
@@ -308,3 +316,208 @@ bool hasInsufficientMaterial(Color c, const Position& pos) {
 
     return true;
 }
+
+enum FEN_VALIDATION : int {
+  FEN_MISSING_SPACE_DELIM = -12,
+  FEN_INVALID_NB_PARTS = -11,
+  FEN_INVALID_CHAR = -10,
+  FEN_FILE_SIZE_MISMATCH = -9,
+  FEN_INVALID_NUMBER_OF_RANKS = -8,
+  FEN_INVALID_POCKET_INFO = -7,
+  FEN_INVALID_SIDE_TO_MOVE = -6,
+  FEN_INVALID_CASTLING_INFO = -5,
+  FEN_INVALID_EN_PASSANT_SQ = -4,
+  FEN_INVALID_CHECK_COUNTER = -3,
+  FEN_INVALID_HALF_MOVE_COUNTER = -2,
+  FEN_INVALID_MOVE_COUTNER = -1,
+  FEN_EMPTY = 0,
+  FEN_OK = 1
+};
+
+
+FEN_VALIDATION validate_fen(const std::string& fen, const Variant* v) {
+
+  // 0) Layout
+  // check for empty fen
+  if (fen.size() == 0) {
+    std::cerr << "Fen is empty." << std::endl;
+    return FEN_EMPTY;
+  }
+
+  // check for space
+  if (fen.find(' ') == std::string::npos) {
+    std::cerr << "Fen misses space as delimiter." << std::endl;
+    return FEN_MISSING_SPACE_DELIM;
+  }
+
+  std::vector<std::string> fenParts;
+  std::vector<std::string> starFenParts;
+  std::stringstream ss(fen);
+  std::stringstream ss2(v->startFen);
+
+  std::string curPart;
+  while (std::getline(ss, curPart, ' '))
+    fenParts.emplace_back(curPart);
+  while (std::getline(ss2, curPart, ' '))
+    starFenParts.emplace_back(curPart);
+  const unsigned int nbFenParts = starFenParts.size();
+
+  // check for number of parts
+  if (fenParts.size() != nbFenParts) {
+    std::cerr << "Invalid number of fen parts. Expected: " << nbFenParts << " Actual: " << fenParts.size() << std::endl;
+    return FEN_INVALID_NB_PARTS;
+  }
+
+  // 1) Part
+  // check for valid characters
+  for (char c : fenParts[0])
+    if (!isdigit(c) && v->startFen.find(c) == std::string::npos) {
+      std::cerr << "Invalid piece character: '" << c << "'." << std::endl;
+      return FEN_INVALID_CHAR;
+    }
+
+  // check for number of ranks
+  int nbRanks = RANK_1;
+  for (; nbRanks <= v->maxRank; ++nbRanks);
+  // check for number of files
+  int nbFiles = FILE_A;
+  for (; nbFiles <= v->maxFile; ++nbFiles);
```

Comment:
These for loops do not seem to be doing much. Does anything speak against just setting them to the target values right away?

Reply:
You're right, setting them directly is better.
```C++
int nbRanks = v->maxRank - RANK_1 + 1;
int nbFiles = v->maxFile - FILE_A + 1;
```

Reply:
> The question is if the validator should only accept the correct FEN or also others that are accepted by the parser, but perhaps the former makes more sense in order to encourage sticking to standards

I also allowed using an additional `/` for specifying the pockets as this is the case on [lichess.org](https://lichess.org/).

Reply:
Now that we talk about it, getting the number of ranks/files for a variant is actually a quite common operation in the code base, and I usually for simplicity just write `v->maxRank + 1`, but subtracting `RANK_1` is of course more correct conceptually. Introducing methods for rank and file count in `Variant` to replace those "count = max +1" usages perhaps makes sense, but you do not need to do it as part of the PR, since it is not really related, the code as you wrote it as in your comment should be fine.